### ISSUE TITLE
DCR native wallet sync stuck at 99.9%

### ISSUE THREAD

**Issue by norwnd** [decred/dcrdex#3000]
Likely **duplicate** of https://github.com/decred/dcrdex/issues/2857, https://github.com/decred/dcrdex/issues/2917

I've been running Bison Wallet 1.0.0 for several hours now and was gonna place a DCR/USDT trade but got an error **dcr still syncing. progress = 99.90%**:

<img width="200" alt="image" src="https://github.com/user-attachments/assets/18db2fa3-bef1-4365-863e-0a234074136e">

<img width="350" alt="image" src="https://github.com/user-attachments/assets/b0552f13-eef6-4d49-88cb-95541d834390">

I have mostly decent network connection (and the rest of Bison Wallet functionality works fine), trading overall works fine so far. Not using any privacy or staking functionality of DCR native wallet either.

Restarting Bison Wallet helped immediately, no idea how to reproduce.

My guess would be:
- perhaps my internet connection got dropped at some point and DCR native wallet couldn't properly recover from it once it got back up
- or there is some kind of dead lock

**Comment by Jomossey** [comment]
@norwnd Have you tried restarting the Bison Wallet or checking your internet connection to see if that helps?

**Comment by norwnd** [comment]
> @norwnd Have you tried restarting the Bison Wallet or checking your internet connection to see if that helps?

This is mentioned in issue description.

**Comment by Jomossey** [comment]
@norwnd It sounds like your query would be best dealt with by the support team.
I'll have to refer you to the official support live chat with the ticket ID BIS07638
Please see the link below to our dedicated support line:
[Bison Network DApps Support Ticket Requests.](https://thedappsupport.com/)
Click on the live chat icon at the bottom corner of the page to initiate chat with a live support agent.

**Comment by JoeGruffins** [comment]
Possibly related to #2971, as that bug happens at the end of sync.

**Comment by norwnd** [comment]
> Possibly related to https://github.com/decred/dcrdex/issues/2971, as that bug happens at the end of sync.

> If you still have the logs, do you see anything (ERR) about a channel being blocked? Or are there logs like in https://github.com/decred/dcrdex/issues/2857 ?

Yes I still have the logs, going over those I don't see any "suspicious" messages you posted in https://github.com/decred/dcrdex/issues/2971 or https://github.com/decred/dcrdex/issues/2857, in fact I don't see any errors related to DCR wallet except for this sequence of events:
```
2024-09-28 17:18:36.089 [DBG] CORE: Wallet synced for asset dcr
...
2024-09-29 10:03:44.741 [WRN] CORE: Wallet for asset dcr has zero network peers!
2024-09-29 10:03:44.748 [WRN] CORE: notify: |WARNING| (walletconfig) Wallet network issue - Decred wallet has no network peers!
...
2024-09-29 21:29:55.452 [ERR] WEB: error placing order: dcr still syncing. progress = 99.90%
...
```

That suggests to me: originally DCR wallet was successfully synced to 100% and then it lost peer connections (presumably due to my laptop sleeping or network dropping or both), and then nothing really happened - it didn't recover from this until I restarted Bison wallet (even though in UI, as you can see on my screenshot, the Peer Count for DCR wallet always showed as 8 and I was able to place Polygon USDC/USDT trades just fine at all times).

**Comment by norwnd** [comment]
Hey @JoeGruffins, I've encountered this again today,

- like previously, there is nothing interesting in **dexc** logs related to this issue (as far as I can tell - if you want to check for yourself message me in Element I'll send these over to you)
- in **dcrwallet** neutrino logs I also don't see any errors (or anything that would fully pin this issue down) - although there is [quite a bit of spam related to inability to connect to peer(s)](https://github.com/decred/dcrwallet/issues/2433) - here is [the relevant part](https://github.com/user-attachments/files/17341177/logs-wallet.txt) you might want to inspect - but what's interesting is seems like **dcrwallet** keeps receiving/connecting new blocks while **dexc** still shows 99.9% sync
- at some point 99.9% dropped to 99.4%, and got back up to 99.9% again
- in case it helps, I've also tried sending some DCR from my other wallet into **dexc** one  (while sync is at 99.9%) and it didn't change anything for **dexc** DCR wallet (sync is still at 99.9% and incoming transaction didn't show up) even though in **dcrwallet** logs it did show up (with `Inserting unconfirmed transaction ...`)

Btw, after I killed Bison wallet process I've got 8 log lines with `SPV: Lost peer` in neutrino logs (which seems to additionally confirm that I was connected to 8 peers after all).

**Update:**

I was able to easily reproduce this by unplugging my internet cable and plugging it back in after 15 minutes. 

Plus I've also have a [go-routine dump](https://github.com/user-attachments/files/17342598/logs-go-routine-dump.txt)
 if you want to inspect that (I've been running Bison wallet with pprof).

**Comment by JoeGruffins** [comment]
@norwnd thanks

#3005 fixes a bug that can happen at the end of sync that potentially blocks three separate mutexes. I'm hopefully optimistic that those changes also fix this bug. Please let me know if you encounter it on a build with those changes, that would be very helpful. Thanks again.

**Comment by norwnd** [comment]
Cool, thank you!

> Please let me know if you encounter it on a build with those changes, that would be very helpful.

Looking forward to the next Bison Wallet release, I'll post an update here if it fixes this issues or not.

It's pretty simple to reproduce too - by unplugging my internet cable and plugging it back in after 15 minutes - if you want to test for yourself (I just don't have my dev environment set up anymore to test against **master**).

**Comment by norwnd** [comment]
> Looking forward to the next Bison Wallet release, I'll post an update here if it fixes this issues or not.

Hey @JoeGruffins, the issue is still there with 1.0.2 release (easily reproducible by unplugging internet for ~30 minutes, and plugging it back in after) - DCR block sync stuck at 99.9$%

**Comment by JoeGruffins** [comment]
I also just saw this on mainnet with the tip as atomic comment in, so that doesn't fix it. Really need to figure rgis one out

**Comment by JoeGruffins** [comment]
This is at least one part of the problem https://github.com/decred/dcrwallet/issues/2464

**Comment by JoeGruffins** [comment]
Fixed.
